src/multivariate/dirichlet.jl

```diff
@@ -375,3 +375,63 @@ function fit_mle(::Type{<:Dirichlet}, P::AbstractMatrix{Float64},
     elogp = mean_logp(suffstats(Dirichlet, P, w))
     fit_dirichlet!(elogp, α; maxiter=maxiter, tol=tol, debug=debug)
 end
+
+## Differentiation
+function ChainRulesCore.frule((_, Δalpha)::Tuple{Any,Any}, DT::Union{Type{Dirichlet{T}}, Type{Dirichlet}}, alpha::AbstractVector{T}; check_args = true) where {T}
+    d = DT(alpha; check_args=check_args)
+    Δalpha = ChainRulesCore.unthunk(Δalpha)
+    ∂alpha0 = sum(Δalpha)
+    ∂lmnB::typeof(∂alpha0) = sum(Δalpha[i] * (SpecialFunctions.digamma(alpha[i]) - SpecialFunctions.digamma(d.alpha0)) for i in eachindex(alpha))
+    backing = (alpha=Δalpha, alpha0=∂alpha0, lmnB=∂lmnB)
+    t = ChainRulesCore.Tangent{typeof(d), NamedTuple{(:alpha, :alpha0, :lmnB), Tuple{typeof(alpha), typeof(d.alpha0), typeof(d.lmnB)}}}(backing)
+    return d, t
+end
+
+function ChainRulesCore.rrule(DT::Union{Type{Dirichlet{T}}, Type{Dirichlet}}, alpha::AbstractVector{T}; check_args = true) where {T}
+    d = DT(alpha; check_args=check_args)
+    function dirichlet_pullback(d_dir)
+        d_dir = ChainRulesCore.unthunk(d_dir)
+        ∂l = d_dir.lmnB * (SpecialFunctions.digamma.(alpha) .- SpecialFunctions.digamma.(d.alpha0))
+        return (ChainRulesCore.NoTangent(), d_dir.alpha .+ d_dir.alpha0 .+ ∂l)
+    end
+    return d, dirichlet_pullback
+end
+
+function ChainRulesCore.frule((_, Δd, Δx), ::typeof(_logpdf), d::Dirichlet, x::AbstractVector{T}) where {T}
+    lp = _logpdf(d, x)
+    if !insupport(d, x)
+        return (lp, zero(lp))
+    end
+    ∂α = sum(Δd.alpha[i] * log(x[i]) for i in eachindex(x))
+    ∂l = - Δd.lmnB
+    ∂x = sum((d.alpha[i] - 1) * Δx[i] / x[i] for i in eachindex(x))
+    return (lp, ∂α + ∂l + ∂x)
+end
+
+function ChainRulesCore.rrule(::typeof(_logpdf), d::Dirichlet, x::AbstractVector{T}) where {T}
+    y = _logpdf(d, x)
+    function Dirichlet_logpdf_pullback(dy)
+        if !isfinite(y)
+            backing = (alpha = zero(d.alpha), alpha0 = ChainRulesCore.ZeroTangent(), lmnB=zero(d.lmnB))
+            ∂d = ChainRulesCore.Tangent{typeof(d), typeof(backing)}(backing)
+            ∂x = zero(d.alpha + x)
+            return (ChainRulesCore.NoTangent(), ∂d, ∂x)
+        end
+        ∂alpha = dy * log.(x)
```

Comment:
Saves allocations:
```suggestion
        ∂alpha = dy .* log.(x)
```
Additionally, maybe for the case `dy = x = 0` it would be useful to use `xlogy`?